ext/spl/spl_rewindablekeyvalueiterator.c

```diff
@@ -0,0 +1,473 @@
+	*count = intern->array.size;
+	return SUCCESS;
+}
+
+PHP_METHOD(RewindableKeyValueIterator, __construct)
+{
+	zval *object = ZEND_THIS;
+	spl_rewindablekeyvalueiterator_object *intern;
+	zval* iterable;
+
+	ZEND_PARSE_PARAMETERS_START(1, 1)
+		Z_PARAM_ITERABLE(iterable)
+	ZEND_PARSE_PARAMETERS_END();
+
+	intern = Z_REWINDABLEKEYVALUEITERATOR_P(object);
+
+	if (!spl_rewindablekeyvalueiterator_empty(&intern->array)) {
+		/* called __construct() twice, bail out */
+		return;
+	}
+
+	intern->offset = 0;
+	switch (Z_TYPE_P(iterable)) {
+		case IS_ARRAY:
+			spl_rewindablekeyvalueiterator_init_from_array(&intern->array, Z_ARRVAL_P(iterable));
+			return;
+		case IS_OBJECT:
+			spl_rewindablekeyvalueiterator_init_from_traversable(&intern->array, Z_OBJ_P(iterable));
+			return;
+		EMPTY_SWITCH_DEFAULT_CASE();
+	}
+}
+
+PHP_METHOD(RewindableKeyValueIterator, count)
+{
+	zval *object = ZEND_THIS;
+	spl_rewindablekeyvalueiterator_object *intern;
+
+	if (zend_parse_parameters_none() == FAILURE) {
+		RETURN_THROWS();
+	}
+
+	intern = Z_REWINDABLEKEYVALUEITERATOR_P(object);
+	RETURN_LONG(intern->array.size);
+}
+
+/* Returns the key at the current offset */
+PHP_METHOD(RewindableKeyValueIterator, key)
+{
+	spl_rewindablekeyvalueiterator_object  *intern;
+
+	if (zend_parse_parameters_none() == FAILURE) {
+		RETURN_THROWS();
+	}
+
+	intern = Z_REWINDABLEKEYVALUEITERATOR_P(ZEND_THIS);
+	if (intern->offset >= intern->array.size) {
+		RETURN_NULL();
+	}
+	RETURN_COPY(&intern->array.entries[intern->offset].key);
+}
+
+/* Returns the value at the current offset */
+PHP_METHOD(RewindableKeyValueIterator, current)
+{
+	spl_rewindablekeyvalueiterator_object  *intern;
+
+	if (zend_parse_parameters_none() == FAILURE) {
+		RETURN_THROWS();
+	}
+
+	intern = Z_REWINDABLEKEYVALUEITERATOR_P(ZEND_THIS);
+	if (intern->offset >= intern->array.size) {
+		RETURN_NULL();
+	}
+	RETURN_COPY(&intern->array.entries[intern->offset].value);
+}
+
+/* Advance the iterator */
+PHP_METHOD(RewindableKeyValueIterator, next)
+{
+	spl_rewindablekeyvalueiterator_object  *intern;
+
+	if (zend_parse_parameters_none() == FAILURE) {
+		RETURN_THROWS();
+	}
+
+	intern = Z_REWINDABLEKEYVALUEITERATOR_P(ZEND_THIS);
+	if (intern->offset < intern->array.size) {
+		intern->offset++;
+	}
+}
+
+/* Rewind the iterator */
+PHP_METHOD(RewindableKeyValueIterator, rewind)
+{
+	spl_rewindablekeyvalueiterator_object  *intern;
+
+	if (zend_parse_parameters_none() == FAILURE) {
+		RETURN_THROWS();
+	}
+
+	intern = Z_REWINDABLEKEYVALUEITERATOR_P(ZEND_THIS);
+	intern->offset = 0;
+}
+
+/* Check if iterator is valid */
+PHP_METHOD(RewindableKeyValueIterator, valid)
+{
+	spl_rewindablekeyvalueiterator_object  *intern;
+
+	if (zend_parse_parameters_none() == FAILURE) {
+		RETURN_THROWS();
+	}
+
+	intern = Z_REWINDABLEKEYVALUEITERATOR_P(ZEND_THIS);
+	RETVAL_BOOL(intern->offset < intern->array.size);
+}
+
+// TODO: Implement internal iterator functions for better performance? See spl_fixedarray.c
```

Comment:
Yes, please do! It avoids calling into the VM and is much faster.

Reply:
Looking again, I can probably do what SplFixedArray was refactored to do in php 8.x-dev and use `\InternalIterator` - `spl_fixedarray_it`